server/templatetags/i18n_icu.py

```diff
@@ -99,9 +99,18 @@ def trans_html(
             tags[placeholder]["attrs"][tag_attr] = kwargs[arg]
             continue
 
+    locale = (
+        # if asked for a specific locale, we use it
+        locale
+        # else, we look for the i18n context and get the value from there
+        or context.get("i18n", {}).get("locale_id")
+        # sometimes (e.g. in `allauth` emails) we don't have the i18n context; in this case, we inspect the current request
```

Comment:
It builds its own context: https://github.com/pennersr/django-allauth/blob/master/allauth/account/forms.py#L531

Reply:
Eep. Confusing. Let's try and enumerate all the cases:

1. Caller passes a `locale` argument to `trans_html()`. (Why is this a valid case? Do we ever use this in production?)
2. Rendering a Django template from Django view. `context` is set by context processors.
3. Rendering a notification email: the caller sets `context`; there is no request.
4. Rendering Django-allauth: Django-allauth overwrites Django behavior so Django doesn't use the context processors we configured ... but luckily, there's a `"request"` key in the contexts allauth passes us.

This seems like an obvious error in Django-allauth. (It's a legacy thing; nowadays, they should be passing request to `render_to_string()`.)

I think we can circumvent the bug. Please write (and configure Django to use) a child of `allauth.account.DefaultAccountAdapter` and override its `send_mail()` to `return super().send_mail(template_prefix, email, RequestContext(context["request"], context)))`. Crash if `context["request"]` does not exist, because that means our assumptions are wrong and we need to be emailed ASAP to fix our bug.

... my goal here is to remove locale-selecting logic. It would be great if there were only one way to select locale. I think the obvious way is for us to always use `context["i18n"]["locale_id"]`. Don't even catch `KeyError` -- that would mean a bug elsewhere, meaning Undefined Behavior ... meaning please crash so we're emailed ASAP and can fix our bug.

Reply:
There is already a child of `allauth.account.DefaultAccountAdapter` in `cjworkbench/views/account_adapter` which looks like it's unused (it has been commented out in `settings.py`). Should I keep it or delete it?

Reply:
WOW. Please delete that file and start over.

(This was my bad. We used to construct different emails in production than in dev/test. On a day of Sendgrid emergency, I couldn't test production-only code and so I commented it out to deploy the tested code. I should have deleted it.)

Reply:
Oh, and.... please don't put the file in `views/`. It never belonged there. Just put it one level above. And maybe include "allauth" in the filename so we don't get confused again -- "AccountAdapter" is a stupidly-generic name.

Reply:
`super().send_mail` was complaining that it needs a `dict` instead of a `RequestContext`, so I just wrote `return super().send_mail(template_prefix, email, {**context, **context_processor(context["request"])})` and it works.

Regarding the `{% trans %}` template tag, in general it is useful to be able to bypass the global setting. However, we haven't encountered any such situation so far in this project and I don't know whether we will.